src/query/functions/linear/histogram_quantile_test.go

```diff
@@ -0,0 +1,283 @@
+		{upperBound: 1, value: 1},
+		{upperBound: 2, value: 2},
+		{upperBound: 5, value: 5},
+		{upperBound: 10, value: 10},
+		{upperBound: 20, value: 15},
+		{upperBound: math.Inf(1), value: 16},
+	}
+
+	actual = bucketQuantile(0, buckets)
+	assert.InDelta(t, float64(0), actual, 0.0001)
+
+	actual = bucketQuantile(0.15, buckets)
+	assert.InDelta(t, 2.4, actual, 0.0001)
+
+	actual = bucketQuantile(0.2, buckets)
+	assert.InDelta(t, float64(3.2), actual, 0.0001)
+
+	actual = bucketQuantile(0.5, buckets)
+	assert.InDelta(t, float64(8), actual, 0.0001)
+
+	actual = bucketQuantile(0.8, buckets)
+	assert.InDelta(t, float64(15.6), actual, 0.0001)
+
+	actual = bucketQuantile(1, buckets)
+	assert.InDelta(t, float64(20), actual, 0.0001)
+}
+
+func TestNewOp(t *testing.T) {
+	args := make([]interface{}, 0, 1)
+	_, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.Error(t, err)
+
+	args = append(args, "invalid")
+	_, err = NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.Error(t, err)
+
+	args[0] = 2.0
+	_, err = NewHistogramQuantileOp(args, ClampMaxType)
+	assert.Error(t, err)
+
+	op, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.NoError(t, err)
+
+	assert.Equal(t, HistogramQuantileType, op.OpType())
+	assert.Equal(t, "type: histogram_quantile", op.String())
+}
+
+func testQuantileFunctionWithQ(t *testing.T, q float64) [][]float64 {
+	args := make([]interface{}, 0, 1)
+	args = append(args, q)
+	op, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	require.NoError(t, err)
+
+	name := []byte("name")
+	bucket := []byte("bucket")
+	tagOpts := models.NewTagOptions().
+		SetIDSchemeType(models.TypeQuoted).
+		SetMetricName(name).
+		SetBucketName(bucket)
+
+	tags := models.NewTags(3, tagOpts).SetName([]byte("foo")).AddTag(models.Tag{
+		Name:  []byte("bar"),
+		Value: []byte("baz"),
+	})
+
+	seriesMetas := []block.SeriesMeta{
+		{Tags: tags.Clone().SetBucket([]byte("1"))},
+		{Tags: tags.Clone().SetBucket([]byte("2"))},
+		{Tags: tags.Clone().SetBucket([]byte("5"))},
+		{Tags: tags.Clone().SetBucket([]byte("10"))},
+		{Tags: tags.Clone().SetBucket([]byte("20"))},
+		{Tags: tags.Clone().SetBucket([]byte("Inf"))},
+		// this series should not be part of the output, since it has no bucket tag.
+		{Tags: tags.Clone()},
+	}
+
+	v := [][]float64{
+		{1, 1, 11, 12, 1},
+		{2, 2, 12, 13, 2},
+		{5, 5, 15, 40, 5},
+		{10, 10, 20, 50, 10},
+		{15, 15, 25, 70, 15},
+		{16, 19, 26, 71, 1},
+	}
+
+	bounds := models.Bounds{
+		Start:    time.Now(),
+		Duration: time.Minute * 5,
+		StepSize: time.Minute,
+	}
+
+	bl := test.NewBlockFromValuesWithSeriesMeta(bounds, seriesMetas, v)
+	c, sink := executor.NewControllerWithSink(parser.NodeID(1))
+	node := op.(histogramQuantileOp).Node(c, transform.Options{})
+	err = node.Process(parser.NodeID(0), bl)
+	require.NoError(t, err)
+
+	return sink.Values
+}
+
+var (
+	inf  = math.Inf(+1)
+	ninf = math.Inf(-1)
+)
+
+func TestQuantileFunctionFilteringWithoutA(t *testing.T) {
```

Comment:
`WithoutA`?

Reply:
Same with `WithQ`?

Reply:
Oh I copied some setup stuff from an agg function that filtered by an `A` tag, will update this name to make more sense